implementations/rust/ockam/ockam_app_lib/src/incoming_services/state.rs

```diff
@@ -0,0 +1,365 @@
+                        node_route: None,
+                        id: "project_id".to_string(),
+                        name: "project_name".to_string(),
+                        identity_id: Some(
+                            Identifier::from_str("I1234561234561234561234561234561234561234")
+                                .unwrap(),
+                        ),
+                        authority: None,
+                        okta: None,
+                    }),
+                    None,
+                )
+                .hex_encoded()
+                .unwrap(),
+            });
+
+            app_state
+                .load_services_from_invites(vec![invitation.clone()])
+                .await;
+            let services = app_state.incoming_services().read().await.services.clone();
+            assert_eq!(1, services.len(), "One service should be loaded");
+
+            let service = &services[0];
+            assert_eq!(
+                "invitation_id",
+                service.id()
+            );
+            assert_eq!(
+                "I12ab34cd56ef12ab34cd56ef12ab34cd56ef12ab",
+                service.access_details.shared_node_identity.to_string()
+            );
+            assert_eq!("remote_service_name", service.name());
+            assert!(service.enabled());
+            assert!(service.port().is_none());
+            assert_eq!("project_id", service.enrollment_ticket().unwrap().project.unwrap().name, "project name should be overwritten with project id");
+            assert_eq!("forward_to_I12ab34cd56ef12ab34cd56ef12ab34cd56ef12ab", service.relay_name());
+            assert_eq!("/project/project_id/service/forward_to_I12ab34cd56ef12ab34cd56ef12ab34cd56ef12ab/secure/api/service/remote_service_name", service.service_route().unwrap());
+            assert_eq!("ockam_app_project_id_invitation_id", service.local_node_name().unwrap());
+
+            let second_invitation = InvitationWithAccess {
+                invitation: ReceivedInvitation {
+                    id: "second_invitation_id".to_string(),
+                    expires_at: "2020-09-12T15:07:14.00".to_string(),
+                    grant_role: RoleInShare::Admin,
+                    owner_email: "owner_email".to_string(),
+                    scope: ShareScope::Project,
+                    target_id: "target_id".to_string(),
+                    ignored: false,
+                },
+                service_access_details: invitation.service_access_details.clone(),
+            };
+
+            // let's load another invitation, but but a persistent state for it already exists
+            app_state.model_mut(|m| {
+                m.incoming_services.push(PersistentIncomingServiceState {
+                    invitation_id: "second_invitation_id".to_string(),
+                    enabled: false,
+                    name: Some("custom_user_name".to_string()),
+                })
+            }).await.unwrap();
+
+            app_state
```

Comment:
One slightly confusing thing with this test is that both services share the same access details. So I wondered if that was an objective of the test, testing what happens when there are 2 invitations to the same service. But actually I think that this is just about testing what happens when some state (service name / service enabled) has already been persisted.

Reply:
Access details are just carried over and wasn't the objective the test, i will try to better explain the test goal by renaming and adding some comments